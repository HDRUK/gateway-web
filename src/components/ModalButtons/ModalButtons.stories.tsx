import type { Meta, StoryObj } from "@storybook/react";
import ModalButtons from "./ModalButtons";

const meta: Meta<typeof ModalButtons> = {
    component: ModalButtons,
};

export default meta;

type Story = StoryObj<typeof ModalButtons>;

const WrapperComponent = () => {
    return null;
};

export const Default: Story = {
    render: () => <WrapperComponent />,
};
