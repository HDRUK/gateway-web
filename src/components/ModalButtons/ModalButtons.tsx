import useDialog from "@/hooks/useDialog";
import Button from "@/components/Button";

type ConfirmType = "button" | "submit" | undefined;
export interface ModalButtonProps {
    onSuccess?: (props: unknown) => void;
    onCancel?: (props: unknown) => void;
    confirmText?: string;
    cancelText?: string;
    confirmType?: ConfirmType;
}

const ModalButtons = ({
    onSuccess,
    onCancel,
    cancelText,
    confirmText,
    confirmType,
}: ModalButtonProps) => {
    const { hideDialog: hideModal } = useDialog();

    const handleSuccess = (props: unknown) => {
        if (typeof onSuccess === "function") {
            onSuccess(props);
        }
        hideModal();
    };

    const handleCancel = (props: unknown) => {
        if (typeof onCancel === "function") {
            onCancel(props);
        }
        hideModal();
    };

    return (
        <>
            <Button
                key="cancel"
                variant="outlined"
                color="secondary"
                onClick={handleCancel}>
                {cancelText}
            </Button>
            {confirmType === "submit" ? (
                <Button key="confirm" type="submit">
                    {confirmText}
                </Button>
            ) : (
                <Button
                    key="confirm"
                    type={confirmType}
                    onClick={handleSuccess}>
                    {confirmText}
                </Button>
            )}
        </>
    );
};

ModalButtons.defaultProps = {
    cancelText: "Cancel",
    confirmText: "Confirm",
    confirmType: "button" as ConfirmType,
    onSuccess: () => null,
    onCancel: () => null,
};

export default ModalButtons;
