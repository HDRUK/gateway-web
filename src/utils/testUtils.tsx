import React, { ReactNode } from "react";
import {
    render,
    renderHook,
    RenderHookOptions,
    RenderOptions,
    RenderResult,
    RenderHookResult,
} from "@testing-library/react";
import { SWRConfig } from "swr";
import { CacheProvider, ThemeProvider } from "@emotion/react";
import ActionBarProvider from "@/providers/ActionBarProvider";
import DialogProvider from "@/providers/DialogProvider";
import theme from "@/config/theme";
import createEmotionCache from "@/config/createEmotionCache";

const clientSideEmotionCache = createEmotionCache();

const Wrapper = ({ children }: { children: ReactNode }) => {
    return (
        <SWRConfig
            value={{
                provider: () => new Map(),
            }}>
            <CacheProvider value={clientSideEmotionCache}>
                <ThemeProvider theme={theme}>
                    <ActionBarProvider>
                        <DialogProvider>{children}</DialogProvider>
                    </ActionBarProvider>
                </ThemeProvider>
            </CacheProvider>
        </SWRConfig>
    );
};

interface OptionProps
    extends RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRender = (
    ui: React.ReactElement<
        unknown,
        string | React.JSXElementConstructor<unknown>
    >,
    options?: OptionProps
): RenderResult => {
    const { wrapperProps, ...rest } = options || {};

    return render(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

interface OptionHookProps
    extends RenderHookOptions<
        unknown,
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRenderHook = (
    ui: (initialProps: unknown) => unknown,
    options?: OptionHookProps
): RenderHookResult<unknown, unknown> => {
    const { wrapperProps, ...rest } = options || {};
    return renderHook(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

export * from "@testing-library/react";

export { customRenderHook as renderHook };
export { customRender as render };
