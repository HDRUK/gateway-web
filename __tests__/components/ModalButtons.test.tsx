import ModalButtons from "@/components/ModalButtons";
import { fireEvent, render, screen, waitFor } from "../testUtils";

describe("ModalButtons", () => {
    it("should render basic modal buttons", () => {
        const wrapper = render(<ModalButtons />);

        expect(wrapper.container).toMatchSnapshot();
    });
    it("should render custom label modal buttons", async () => {
        render(<ModalButtons cancelText="Dismiss" confirmText="Save" />);

        expect(screen.getByText("Dismiss")).toBeInTheDocument();
        expect(screen.getByText("Save")).toBeInTheDocument();
    });
    it("should call custom cancel function", async () => {
        const cancelFn = jest.fn();
        render(<ModalButtons onCancel={cancelFn} />);

        fireEvent.click(screen.getByText("Cancel"));

        await waitFor(() => {
            expect(cancelFn).toHaveBeenCalled();
        });
    });
    it("should call custom confirm function", async () => {
        const confirmFn = jest.fn();
        render(<ModalButtons onSuccess={confirmFn} />);

        fireEvent.click(screen.getByText("Confirm"));

        await waitFor(() => {
            expect(confirmFn).toHaveBeenCalled();
        });
    });
});
