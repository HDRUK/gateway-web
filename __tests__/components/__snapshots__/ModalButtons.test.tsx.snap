// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalButtons should render basic modal buttons 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium mui-style-15l9l0w-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Cancel
    <span
      class="MuiTouchRipple-root mui-style-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium mui-style-sfwpz-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Confirm
    <span
      class="MuiTouchRipple-root mui-style-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
